src/puppetlabs/puppetdb/queue.clj

```diff
@@ -190,47 +199,65 @@
    :certname s/Str
    :producer-ts (s/maybe pls/Timestamp)
    :callback (s/=> s/Any s/Any)
-   :command-stream java.io.InputStream})
+   :command-stream java.io.InputStream
+   :compression compression-file-extension-schema})
 
 (s/defn create-command-req :- command-req-schema
   "Validating constructor function for command requests"
   [command :- s/Str
    version :- s/Int
    certname :- s/Str
    producer-ts :- (s/maybe s/Str)
+   compression :- compression-file-extension-schema
    callback :- (s/=> s/Any s/Any)
    command-stream :- java.io.InputStream]
   {:command command
    :version version
    :certname certname
    :producer-ts (when producer-ts
                   (pdbtime/from-string producer-ts))
+   :compression compression
    :callback callback
    :command-stream command-stream})
 
-(defrecord CommandRef [id command version certname received producer-ts callback delete?])
+(defrecord CommandRef [id command version certname received producer-ts callback delete? compression])
 
 (defn cmdref->entry [{:keys [id received] :as cmdref}]
   (stock/entry id (serialize-metadata received cmdref)))
 
 (defn entry->cmdref [entry]
-  (let [{:keys [received command version certname producer-ts]} (-> entry
-                                                                    stock/entry-meta
-                                                                    parse-metadata)]
+  (let [{:keys [received command version
+                certname producer-ts compression]} (-> entry
+                                                       stock/entry-meta
+                                                       parse-metadata)]
     (map->CommandRef {:id (stock/entry-id entry)
                       :command command
                       :version version
                       :certname certname
                       :received received
                       :producer-ts producer-ts
-                      :callback identity})))
+                      :callback identity
+                      :compression compression})))
+
+(defn wrap-decompression-stream
+  [file-extension command-stream]
+  (case file-extension
+    nil command-stream
+    "" command-stream
+    "gz" (GzipCompressorInputStream. command-stream)
+    (throw+ {:kind ::parse-error} nil
+            (trs "Unsupported compression file extension for command: {0}"
+                 file-extension))))
 
 (defn cmdref->cmd [q cmdref]
-  (let [entry (cmdref->entry cmdref)]
+  (let [compression (:compression cmdref)
+        entry (cmdref->entry cmdref)]
     (with-open [command-stream (stock/stream q entry)]
       (assoc cmdref
-             :payload (stream->json command-stream)
-             :entry entry))))
+        :payload (stream->json (wrap-decompression-stream
```

Comment:
👍 

I was trying to refresh my memory on the order of compression with java streams. I knew there was a big different depending on how it's arranged so I ran some local tests. This is the fastest way to do it. I'm not sure why, but if instead you have the decompression on the result of the buffered stream, it's super slow (i.e. like 400x slower). Might be good to have a comment on the function above to that affect to let people in the future know that order is important here.